pytorch_lightning/utilities/cli.py

```diff
@@ -0,0 +1,245 @@
+        return self.add_class_arguments(lightning_class, nested_key)
+
+
+class SaveConfigCallback(Callback):
+    """Saves a LightningCLI config to the log_dir when training starts"""
+
+    def __init__(self, parser, config):
+        self.parser = parser
+        self.config = config
+
+    def on_train_start(self, trainer, pl_module):
+        config_path = os.path.join(trainer.logger.log_dir, 'config.yaml')
+        self.parser.save(self.config, config_path, skip_none=False)
+
+
+class LightningCLI:
+
+    def __init__(
+        self,
+        model_class: Type[LightningModule],
+        datamodule_class: Type[LightningDataModule] = None,
+        save_config_callback: Type[SaveConfigCallback] = SaveConfigCallback,
+        trainer_class: Type[Trainer] = Trainer,
+        trainer_kwargs: Dict[str, Any] = None,
+        description: str = 'pytorch-lightning trainer command line tool',
+        env_prefix: str = 'PL',
+        env_parse: bool = False,
+        parser_kwargs: Dict[str, Any] = None,
+        subclass_mode_model: bool = False,
+        subclass_mode_data: bool = False
+    ):
+        """
+        Implementation of a configurable command line tool for pytorch-lightning
+
+        Receives as input pytorch-lightning classes, which are instantiated
+        using a parsed configuration file and/or command line args and then runs
+        trainer.fit. Parsing of configuration from environment variables can
+        be enabled by setting :code:`env_parse=True`. A full configuration yaml would
+        be parsed from :code:`PL_CONFIG` if set. Individual settings are so parsed from
+        variables named for example :code:`PL_TRAINER__MAX_EPOCHS`.
+
+        Example, first implement the trainer.py tool as::
+
+            from mymodels import MyModel
+            from pytorch_lightning.utilities.cli import LightningCLI
+            LightningCLI(MyModel)
+
+        Then in a shell, run the tool with the desired configuration::
+
+            $ python trainer.py --print_config > config.yaml
+            $ nano config.yaml  # modify the config as desired
+            $ python trainer.py --cfg config.yaml
+
+        Args:
+            model_class: The LightningModule class to train on.
+            datamodule_class: An optional LightningDataModule class.
+            save_config_callback: A callback class to save the training config.
+            trainer_class: An optional extension of the Trainer class.
+            trainer_kwargs: Additional arguments to instantiate Trainer.
+            description: Description of the tool shown when running --help.
+            env_prefix: Prefix for environment variables.
+            env_parse: Whether environment variable parsing is enabled.
+            parser_kwargs: Additional arguments to instantiate LightningArgumentParser.
+            subclass_mode_model: Whether model can be any `subclass
+                <https://omni-us.github.io/jsonargparse/#classes-as-type>`_ of the
```

Comment:
All the links have been moved and need to be updated

Reply:
Fixed